v0/destinations/hs/HSTransform-v2.js

```diff
@@ -0,0 +1,486 @@
+        );
+
+        // duplicate email can cause issue with create in batch
+        // updating the existing one to avoid duplicate
+        // as same event can fire in batch one of the reason
+        // can be due to network lag or processor being busy
+        const isDuplicate = identifyResponseList.find(data => {
+          return (
+            data.properties.email === ev.message.body.JSON.properties.email
+          );
+        });
+        if (isDefinedAndNotNullAndNotEmpty(isDuplicate)) {
+          // array is being shallow copied hence changes are affecting the original reference
+          // basically rewriting the same value to avoid duplicate entry
+          isDuplicate.properties = ev.message.body.JSON.properties;
+        } else {
+          // appending unique events
+          identifyResponseList.push({
+            properties: ev.message.body.JSON.properties
+          });
+        }
+        metadata.push(ev.metadata);
+      });
+    } else if (batchOperation === "updateContacts") {
+      // update operation
+      chunk.forEach(ev => {
+        // eslint-disable-next-line no-param-reassign
+        ev.message.body.JSON.properties = formatPropertyValueForCRM(
+          ev.message.body.JSON.properties
+        );
+        // update has contactId and properties
+        // extract contactId from the end of the endpoint
+        const id = ev.message.endpoint.split("/").pop();
+
+        // duplicate contactId is not allowed in batch
+        // updating the existing one to avoid duplicate
+        // as same event can fire in batch one of the reason
+        // can be due to network lag or processor being busy
+        const isDuplicate = identifyResponseList.find(data => {
+          return data.id === id;
+        });
+        if (isDefinedAndNotNullAndNotEmpty(isDuplicate)) {
+          // rewriting the same value to avoid duplicate entry
+          isDuplicate.properties = ev.message.body.JSON.properties;
+        } else {
+          // appending unique events
+          identifyResponseList.push({
+            id,
+            properties: ev.message.body.JSON.properties
+          });
+        }
+        metadata.push(ev.metadata);
+      });
+    } else if (batchOperation === "general") {
+      // general identify event
+
+      /* Note: */
+      // 1. for now it is just HS CRM custom objects however later it can be
+      // refactored to accomodate upcoming endpoint and make necessary changes
+      // 2. this operation for now is just being used by rETL.
+
+      chunk.forEach(ev => {
+        // if source is of rETL
+        if (ev.message.source === "rETL") {
+          identifyResponseList.push({ ...ev.message.body.JSON });
+          metadata.push(ev.metadata);
+        }
+      });
+
+      // CRM_CREATE_CUSTOM_OBJECTS has objectType value
+      // extract objectType from the end of the endpoint
+      const objectType = chunk[0].message.endpoint.split("/").pop();
```

Comment:
can we not get this from payload rather than from splitting url like in other places we are doing

Reply:
This field (contactID) just goes in the endpoint only